### PR TITLE
Maintain selection on file/dir deletion in project panel

### LINKED ISSUE #20444: Wrong element focused when deleting items in project panel
### Check for existing issues

- [X] Completed

### Describe the bug / provide steps to reproduce it

After removing a file/folder from sidebar, keyboard focus is switched to the first entry in the panel. This makes deleting multiple files very impractical.

Solution would be to focus the next file after the deleted one.

### Environment

Zed: v0.160.7 (Zed)
OS: Linux X11 arch unknown
Memory: 62.5 GiB
Architecture: x86_64
GPU: NVIDIA GeForce RTX 2060 with Max-Q Design || NVIDIA || 565.57.01

### If applicable, add mockups / screenshots to help explain present your vision of the feature

_No response_

### If applicable, attach your Zed.log file to this issue.

_No response_

### PR DESCRIPTION
**Pull request by smitbarmase**
Closes #20444

- Focus on next file/dir on deletion.
- Focus on prev file/dir in case where it's last item in worktree.
- Tested when multiple files/dirs are being deleted.

Release Notes:

- Maintain selection on file/dir deletion in project panel.

### CODE REVIEW

**Review by SomeoneToIgnore — CHANGES_REQUESTED** [review]
We need more tests added below, please.

File: crates/project_panel/src/project_panel.rs [hunk 1]
```diff
@@ -1274,6 +1293,13 @@ impl ProjectPanel {
                     })??
                     .await?;
                 }
+                if let Some(next_selection) = next_selection {
+                    this.update(&mut cx, |this, cx| {
+                        this.selection = Some(next_selection);
+                        this.autoscroll(cx);
+                        cx.notify();
```
Comment (SomeoneToIgnore) [inline crates/project_panel/src/project_panel.rs:1300]:
```suggestion
```

That seems to be done in the `autoscroll`.

Optional: I personally find `this` less readable than the real name of the variable, so we can change it along the way if you think so too.

File: crates/project_panel/src/project_panel.rs [hunk 2]
```diff
@@ -6325,6 +6350,205 @@ mod tests {
...
+
+        let project = Project::test(fs.clone(), ["/src".as_ref()], cx).await;
+        let workspace = cx.add_window(|cx| Workspace::test_new(project.clone(), cx));
+        let cx = &mut VisualTestContext::from_window(*workspace, cx);
+        let panel = workspace.update(cx, ProjectPanel::new).unwrap();
+
+        // Delete middle file (second.rs) - should select next file (third.rs)
+        toggle_expand_dir(&panel, "src/test", cx);
+        select_path(&panel, "src/test/second.rs", cx);
+        submit_deletion(&panel, cx);
```
Comment (SomeoneToIgnore) [inline crates/project_panel/src/project_panel.rs:6378]: Let's, before every change selection (initial state like here, selection changed after some deletion assert, etc.), assert the way how the tree looked with the selection.
So, there's always "selection before" vs. "selection after" comparison, which makes these tests more readable.

For the tests below too.

File: crates/project_panel/src/project_panel.rs [hunk 3]
```diff
@@ -6325,6 +6350,205 @@ mod tests {
...
+            }),
+        )
+        .await;
+
+        let project = Project::test(fs.clone(), ["/src".as_ref()], cx).await;
+        let workspace = cx.add_window(|cx| Workspace::test_new(project.clone(), cx));
+        let cx = &mut VisualTestContext::from_window(*workspace, cx);
+        let panel = workspace.update(cx, ProjectPanel::new).unwrap();
+
+        // Delete middle file (second.rs) - should select next file (third.rs)
```
Comment (SomeoneToIgnore) [inline crates/project_panel/src/project_panel.rs:6375]: Some of the tests seems somewhat repetitive, e.g. here it had been asserted in the old tests that was adjusted above.

I would prefer if we concentrate on more multi-select edge-cased ones:
* all files & directories are selected and removed
* a directory + first file in the same directory are selected, more files are in the same selected directory + more files below; then all selected entries are removed.
* multiple files selected in different directories and deleted; some have next file, the very last dir with the selection does not
* maybe, something else? dirs in subdirs selected + removed?

**Review by SomeoneToIgnore — COMMENTED** [review]
Thanks, looks good overall, but let's test it a bit better.

**Comment by 0xtimsb** [comment]
@SomeoneToIgnore upon write test cases i realised what i did earlier, i.e. focusing on just next entry (idx + 1), is straight up wrong as say on deleting dir,  it might choose it's child as `next_selection`.

For this to work,  `find_next_selection` fun, should be tree-aware. I updated the logic (trying to mimick functionality of how VSCode does it and what feels sensible).

Now, it chooses bottom-most marked entry which does not have any marked ancestor. And, select it's sibling.

Handled below test cases:
1. middle file removed from directory
2. last file removed from directory
3. root level file removed
4. nested dir + parent dir removed together
5. mixed selection removed (subdir file + parent file + root file)
6. all contents removed (all root dirs + all root files)
7. nested selection removed (parent dir + subdir + file inside subdir)

Note: I had to supress error `no such entry` on delete file (if file does not exists), as it used to panic that thread, and below logic for `next_selection`  wouldn't run. This panic often happens when say you have list like `dir1` and `dir1/a.rs`, so after deleting `dir1` , `a.rs` won't be there.

**Comment by 0xtimsb** [comment]
Let me know if anything could have been done better ^~^

File: crates/project_panel/src/project_panel.rs [hunk 4]
```diff
@@ -1259,28 +1259,113 @@ impl ProjectPanel {
...
+                    }
+                })?;
                 Result::<(), anyhow::Error>::Ok(())
             })
             .detach_and_log_err(cx);
             Some(())
         });
     }
 
+    fn find_next_selection_after_deletion(
```
Comment (SomeoneToIgnore) [inline crates/project_panel/src/project_panel.rs:1298]: Overall, sounds somewhat heavy as we're iterating over all `entries` in the project panel.
That might be thousands of files easily.

Can we be smarter?
We have a `marked_entries()` method that will return all entries that will be deleted.

We could get all their entries in ~O(1) each and 
1. Do the sanitization I've been talking about before, removing all entries that are subfiles/subdirs (this subset can be determined way above this, in the removal method)
2. Then, we can start looking for entries that are adjacent to the last removed entry: presumably, we can find a parent dir, and then reverse-iterate this subset only, finding the new selection target.

File: crates/project_panel/src/project_panel.rs [hunk 5]
```diff
@@ -1259,28 +1259,113 @@ impl ProjectPanel {
...
+        &self,
+        cx: &mut ViewContext<Self>,
+    ) -> Option<SelectedEntry> {
+        let selection = self.selection?;
+        let (worktree_ix, _, _) = self.index_for_selection(selection)?;
```
Comment (SomeoneToIgnore) [inline crates/project_panel/src/project_panel.rs:1303]: Is this even correct? 
We do iterate over a single worktree entries, but `self.marked_entries` can contain entries from different worktrees, so we need to iterate every entry of every worktree in your case?

Imagining a unit test, where `self.selected_entry` belongs to the worktree that's deleted entirely and a few marked entries belong to the other worktree.
This will fail, would not it?

File: crates/project_panel/src/project_panel.rs [hunk 6]
```diff
@@ -6325,6 +6413,399 @@ mod tests {
...
+                "          file3.txt",
+                "          file4.txt",
+                "      file5.txt",
+                "      file6.txt",
+            ],
+            "Should select next file after deleting middle file"
+        );
+
+        // Test Case 2: Delete last file in directory
+        select_path(&panel, "root/dir1/file2.txt", cx);
```
Comment (SomeoneToIgnore) [inline crates/project_panel/src/project_panel.rs:6487]: But that's already selected?

File: crates/project_panel/src/project_panel.rs [hunk 7]
```diff
@@ -6325,6 +6413,399 @@ mod tests {
...
+        );
+
+        submit_deletion(&panel, cx);
+        assert_eq!(
+            visible_entries_as_strings(&panel, 0..15, cx),
+            &[
+                "v root",
+                "    v dir1",
+                "        > subdir1",
+                "    v dir2  <== selected",
```
Comment (SomeoneToIgnore) [inline crates/project_panel/src/project_panel.rs:6512]: So, after removing the last file in the directory, we jump to its neighbour.
I wonder how intuitive that is? We can select the parent entry always, this will simplify the location logic, would not it?

**Review by SomeoneToIgnore — COMMENTED** [review]
Thanks, but this starts to get even more complex after thinking on the new version more.

Guess it will last for a while before we arrive to a "safe" version.
I've marked a few style comments as NIT, those are not very important.

The rest is very important, as there seems to be a logic error, a potential panic, and a perf degradation added with this change.

File: crates/project_panel/src/project_panel.rs [hunk 8]
```diff
@@ -1259,28 +1259,113 @@ impl ProjectPanel {
...
-                            .ok_or_else(|| anyhow!("no such entry"))
-                    })??
-                    .await?;
+                    if let Some(task) = panel
+                        .update(&mut cx, |panel, cx| {
+                            panel
+                                .project
+                                .update(cx, |project, cx| project.delete_entry(entry_id, trash, cx))
+                        })
+                        .ok()
```
[the remark this answers is not in the document]
Comment (0xtimsb) [reply crates/project_panel/src/project_panel.rs:1276]: Thanks, no worries, I'll pick it up. Let's get this right. I like the sanitization approach, as it solves this as well as removes the need to calculate `last_marked_entry_without_marked_ancestors`. It would be just be the last marked entry, given the sanitization list is sorted by its visibility order. I need to figure out this sorting part without using `visible_entries`.

In `last_marked_entry_without_marked_ancestors` logic, I used `visible_entries` for this reason only, as `marked_entries` is just a map, and it was essential to traverse those entries in that order. But, you are right it can go over thousands easily, so I will just way to sort `marked_entries` in that order.

File: crates/project_panel/src/project_panel.rs (hunk 4 above)
Comment (0xtimsb) [reply crates/project_panel/src/project_panel.rs:1298]: Addressed above.

File: crates/project_panel/src/project_panel.rs (hunk 5 above)
Comment (0xtimsb) [reply crates/project_panel/src/project_panel.rs:1303]: My bad. I wasn't aware that one project panel could contain multiple projects (i.e., multiple worktrees). I discovered the "Add Folder to Project..." button upon reading your comment. 

My initial impression of worktrees was that one project panel could only contain one worktree and worktrees are refering to other window's worktree which we don't care about. So, yes this code will lead to incorrect results. I will add a test cases for multiple worktrees now that I have this new knowledge.

Going forward, I will ask more questions, even if they seem obvious at times.

File: crates/project_panel/src/project_panel.rs [hunk 9]
```diff
@@ -1259,28 +1259,113 @@ impl ProjectPanel {
...
+
+        let mut last_marked_entry_without_marked_ancestors: Option<SelectedEntry> = None;
+        for entry in entries {
+            let current_selection = SelectedEntry {
+                worktree_id: *worktree_id,
+                entry_id: entry.id,
+            };
+
+            if !self.marked_entries.contains(&current_selection)
+                && self.selection != Some(current_selection)
```
[the remark this answers is not in the document]
Comment (0xtimsb) [reply crates/project_panel/src/project_panel.rs:1314]: Addressed above reason to iterate `visible_entries`. Will figure out way to sort `marked_entries` as per visiblity or do we even need sorting to get last (bottom-most) sanitized marked item. Let me think.

File: crates/project_panel/src/project_panel.rs [hunk 10]
```diff
@@ -1259,28 +1259,113 @@ impl ProjectPanel {
...
+            {
+                continue;
+            }
+
+            if let Some(last_entry) = &last_marked_entry_without_marked_ancestors {
+                if let Some(last_path) = self
```
[the remark this answers is not in the document]
Comment (0xtimsb) [reply crates/project_panel/src/project_panel.rs:1320]: I see it now. Yes, better to avoid it.

File: crates/project_panel/src/project_panel.rs (hunk 6 above)
Comment (0xtimsb) [reply crates/project_panel/src/project_panel.rs:6487]: Just kept it so that these test cases remain independent. Removing test case should not affect later ones.

File: crates/project_panel/src/project_panel.rs (hunk 7 above)
Comment (0xtimsb) [reply crates/project_panel/src/project_panel.rs:6512]: I used this logic from vscode. I don't have strong opinion on this one but we can easily do it as parent selection. I will go through both, check which makes most sense.

File: crates/project_panel/src/project_panel.rs [hunk 11]
```diff
@@ -1259,28 +1259,113 @@ impl ProjectPanel {
...
+                .project
+                .read(cx)
+                .worktree_for_id(*worktree_id, cx)?
+                .read(cx);
+
+            let entry = worktree.entry_for_id(last_entry.entry_id)?;
+            let entry_ix = entries.iter().position(|e| e.id == entry.id)?;
+
+            // Use next visible entry after last_entry
+            let next_entry = entries[entry_ix + 1..]
```
[the remark this answers is not in the document]
Comment (0xtimsb) [reply crates/project_panel/src/project_panel.rs:1347]: I wonder why it didn't fail this test. Will check before handling out of bound. Just curious.

```rs
    assert_eq!(
            visible_entries_as_strings(&panel, 0..20, cx),
            &[
                "v root",
                "    v dir1  <== marked",
                "        v subdir1",
                "              a.txt",
                "              b.txt",
                "          file1.txt",
                "    v dir2  <== marked",
                "        v subdir2",
                "              c.txt",
                "              d.txt",
                "          file2.txt",
                "      file3.txt  <== marked",
                "      file4.txt  <== selected  <== marked",
            ],
            "State before deleting all contents"
        );

        submit_deletion(&panel, cx);
        assert_eq!(
            visible_entries_as_strings(&panel, 0..20, cx),
            &["v root  <== selected"],
            "Only empty root directory should remain after deleting all contents"
        );
```

**Comment by 0xtimsb** [comment]
Updated to a sanitization approach and added tests for multiple worktrees.

I also had to update the `compare_paths` function in `paths.rs` to handle cases where file names are the same, but the extensions differ. Previously, this compare func completely ignored the presence of extensions when sorting, leading to unpredictable sorting of such files. Since we use stable sorting everywhere, the original order of those files was retained.

This was problematic when choosing the last marked element. If multiple files with the same name were marked in the directory, it could end up selecting the wrong one than what's in `visible_entries`.

Let me know your thoughts on this. I also added a test case for this change, and all existing cases passed.

**Comment by 0xtimsb** [comment]
Let me know if you spot any other nitpicks or performance stuff. I tried to avoid those issues you mentioned in this one.

File: crates/project_panel/src/project_panel.rs [hunk 12]
```diff
@@ -1833,6 +1932,50 @@ impl ProjectPanel {
...
+                });
+
+        let mut result = BTreeSet::new();
+        for (worktree_id, entries) in entries_by_worktree {
+            if let Some(worktree) = project.worktree_for_id(worktree_id, cx) {
+                let worktree = worktree.read(cx);
+
+                let marked_paths: HashSet<Arc<Path>> = entries
+                    .iter()
+                    .filter_map(|entry| worktree.entry_for_id(entry.entry_id))
```
Comment (SomeoneToIgnore) [inline crates/project_panel/src/project_panel.rs:1957]: My spider-sense hint that we can do better and this is dangerously close to something quadratic.
I do not have good ideas on how to fix this, but a few obvious things:
* we do not need add any files here, only dirs can exclude something
* then we can ask during filtering, if filtered entry's path `starts_with && not equal` an entry from `marked_paths`
* I wonder if ordering these `marked_paths` makes any sense, as this way we'll get the "starts_with" triggering faster (if ordered right)
* NIT, but naming could be improved: `marked_paths` (those are something else), `result` (more like `sanitized_paths`) and even `sanitized_marked_entries` (more like `disjoint_paths_for_removal`) could be named better
* I feel there's a better way to check such things in general, but not insisting

File: crates/project_panel/src/project_panel.rs [hunk 13]
```diff
@@ -6325,6 +6471,598 @@ mod tests {
...
+                "          file3.txt",
+                "          file4.txt",
+                "      file5.txt",
+                "      file6.txt",
+            ],
+            "Should select next directory when last file is deleted"
+        );
+
+        // Test Case 3: Delete root level file
+        select_path(&panel, "root/file5.txt", cx);
```
Comment (SomeoneToIgnore) [inline crates/project_panel/src/project_panel.rs:6563]:
```suggestion
        select_path(&panel, "root/file6.txt", cx);
```

Again, this deleting the file with alive neighbours was tested already, let's make this a bit more dangerous.

File: crates/project_panel/src/project_panel.rs [hunk 14]
```diff
@@ -6325,6 +6471,598 @@ mod tests {
...
+                "v root2",
+                "      file6.txt  <== selected  <== marked",
+            ],
+            "State with all remaining files marked"
+        );
+
+        submit_deletion(&panel, cx);
+        assert_eq!(
+            visible_entries_as_strings(&panel, 0..20, cx),
+            &["v root1", "    v dir2", "v root2  <== selected"],
```
Comment (SomeoneToIgnore) [inline crates/project_panel/src/project_panel.rs:6987]: Can we remove a worktree root?
Not insisting to fix that if we cannot, but curious if there's an easy test to add just after this.
Or maybe even delete both roots? Maybe even with marked files in them?

File: crates/project_panel/src/project_panel.rs [hunk 15]
```diff
@@ -1259,28 +1260,126 @@ impl ProjectPanel {
...
+    ) -> Option<SelectedEntry> {
+        if sanitized_entries.is_empty() {
+            return None;
+        }
+
+        let project = self.project.read(cx);
+        let worktree_map: HashMap<WorktreeId, &Worktree> = sanitized_entries
+            .iter()
+            .map(|entry| entry.worktree_id)
+            .collect::<HashSet<WorktreeId>>()
```
Comment (SomeoneToIgnore) [inline crates/project_panel/src/project_panel.rs:1308]:
```suggestion
```

??????

File: crates/project_panel/src/project_panel.rs [hunk 16]
```diff
@@ -1259,28 +1260,126 @@ impl ProjectPanel {
...
+            .into_iter()
+            .filter_map(|id| project.worktree_for_id(id, cx).map(|w| (id, w.read(cx))))
+            .collect();
+        let (worktree_id, worktree) = worktree_map
+            .iter()
+            .max_by(|(_, a), (_, b)| a.root_name().cmp(b.root_name()))
```
Comment (SomeoneToIgnore) [inline crates/project_panel/src/project_panel.rs:1314]: Again, what's up with all these intermediate single-use allocations?
`.max_by` is in the iterator methods already.

File: crates/project_panel/src/project_panel.rs [hunk 17]
```diff
@@ -1259,28 +1260,126 @@ impl ProjectPanel {
...
+            .map(|(id, worktree)| (*id, *worktree))?;
+
+        let marked_entries_in_worktree: Vec<&SelectedEntry> = sanitized_entries
+            .iter()
+            .filter(|e| e.worktree_id == worktree_id)
+            .collect();
```
Comment (SomeoneToIgnore) [inline crates/project_panel/src/project_panel.rs:1320]:
```suggestion

```
And again...

File: crates/project_panel/src/project_panel.rs [hunk 18]
```diff
@@ -1259,28 +1260,126 @@ impl ProjectPanel {
...
+                    })
+            })
+            .cloned()
+            .collect();
+
+        project::sort_worktree_entries(&mut siblings);
+        let entry_index = siblings.iter().position(|sibling| sibling.id == entry.id)?;
+
+        // Try next sibling
+        if entry_index + 1 < siblings.len() {
```
Comment (SomeoneToIgnore) [inline crates/project_panel/src/project_panel.rs:1357]: This is not needed given that you `.get` right below.

**Review by SomeoneToIgnore — COMMENTED** [review]
This looks better, at least we're definitely not doing anything too risky in terms of panics and iterations (that paths sanitation thing is somewhat close, though).

File: crates/project_panel/src/project_panel.rs (hunk 14 above)
Comment (0xtimsb) [reply crates/project_panel/src/project_panel.rs:6993]: I tried this, but there seems to be an issue with the existing deletion logic when the root is selected or marked.

E.g.
```
> root1
   > some.txt
> root2
```

Case 1: If you select `some.txt` first and then mark `root2`, it will prompt for the deletion of `some.txt`, which is expected.

Case 2: If you select `root2` first and then mark `some.txt`, it will not prompt for anything. It should prompt for `some.txt`.

I'm considering fixing this in another PR, as it might require changes to either the `marked_entries` function (which is being used at lot places) or the logic for marking/selecting the root. I haven’t fully dug into this yet.

Comment (SomeoneToIgnore) [reply crates/project_panel/src/project_panel.rs:6993]: Yeah, we're quite stuck here anyway, so let's not touch the roots yet, thank you for considering this.

File: crates/project_panel/src/project_panel.rs (hunk 17 above)
Comment (0xtimsb) [reply crates/project_panel/src/project_panel.rs:1315]: I converted this into a set, so collect is necessary here. While iterating over siblings below, we need to exclude already marked entries, and I use this set for that check.

**Comment by 0xtimsb** [comment]
I've added comments where I need your input. Everything else is resolved.

Thank you for your patience! I'm still getting the hang of Rust (it's been 3 weeks) and trying not to overlook time complexities. I appreciate it.

**Review by SomeoneToIgnore — APPROVED** [review]
I've pushed a small cleanup pass, but nothing clever or global, just style nits and fewer allocations.

Thank you for bearing up to here, and those tests are very useful!
